### PR TITLE
Cache aspect writers

### PR DESCRIPTION
**Pull request by shriramshankar**

### CODE REVIEW

**Comment by shriramshankar** [comment]
There are 4 parts to this PR and each of the part is in its own commit.

File: cache/sampleStore.js [hunk 1]
```diff
@@ -112,26 +112,28 @@ function removeNullsAndStringifyArrays(obj, arrayFields) {
...
-  for (let j = constants.ISOfields.length - 1; j >= 0; j--) {
-    key = constants.ISOfields[j];
-    obj[key] = obj[key] ? obj[key].toISOString() :
-      new Date().toISOString();
-  }
-
+  constants.ISOfields.forEach((field) => {
+    if (!obj[field]) {
+      obj[field] = new Date().toISOString();
+    } else if (obj[field] && obj[field].toISOString) {
```
Comment (pallavi2209) [inline cache/sampleStore.js]: obj[field].toISOString()

File: api/v1/controllers/aspects.js [hunk 2]
```diff
@@ -175,10 +174,38 @@ module.exports = {
...
+              });
+              return redisOps.hmSet('aspect', cachedAspect.name, cachedAspect);
+            }
+
+            return Promise.resolve(true);
+          })
+          .catch((err) => Promise.resolve(err));
+      }
+
+      return Promise.resolve(true);
```
Comment (pallavi2209) [inline api/v1/controllers/aspects.js]: Some comments would be a plus!

File: cache/redisOps.js [hunk 3]
```diff
@@ -46,6 +47,23 @@ function hmSet(objectName, name, value) {
 } // hmSet
 
 /**
+ * Get the value that is mapped to a key
+ * @param  {String} type - The type of the object on which the operation is to
+ * be performed
+ * @param  {String} name -  Name of the key
+ * @returns {Promise} - which resolves to the value associated with the key
+ */
+function getValue(type, name) {
```
Comment (pallavi2209) [inline cache/redisOps.js]: There is already a function getHashPromise with same intent. This method converts arrayStringsToJson as well. Any way we can avoid code duplication? If you insist on keeping the function, we should atleast change the name and comments to reflect how this function is different from other.

File: api/v1/controllers/aspects.js [hunk 4]
```diff
@@ -237,7 +264,26 @@ module.exports = {
...
+          }
+
+          return Promise.resolve(true);
+        })
+        .then(() => doDeleteAllAssoc(req, res, next, helper,
+              helper.belongsToManyAssoc.users))
+        .catch((err) => u.handleError(next, err, helper.modelName));
+    } else {
+      doDeleteAllAssoc(req, res, next, helper, helper.belongsToManyAssoc.users);
+    }
```
Comment (pallavi2209) [inline api/v1/controllers/aspects.js]: Some comments if you can.

File: api/v1/controllers/aspects.js [hunk 5]
```diff
@@ -251,8 +297,44 @@ module.exports = {
...
+              .filter((writer) => writer !== userName);
+          return redisOps.hmSet('aspect', cachedAspect.name, cachedAspect);
+        }
+
+        return Promise.resolve(true);
+      })
+      .then(() => doDeleteOneAssoc(req, res, next, helper,
+        helper.belongsToManyAssoc.users, userNameOrId))
+      .catch((err) => u.handleError(next, err, helper.modelName));
+    } else {
```
Comment (pallavi2209) [inline api/v1/controllers/aspects.js]: Inline Comments or update function comment please!

Comment (shriramshankar) [inline api/v1/controllers/aspects.js]: What comments do you need ??

File: cache/sampleStore.js (hunk 1 above)
Comment (shriramshankar) [inline cache/sampleStore.js]: what do you mean ?

Comment (pallavi2209) [inline cache/sampleStore.js]: Nevermind, I just checked obj[field].toISOString will return undefined for something other than Date.

File: tests/cache/sampleStore.js [hunk 6]
```diff
@@ -178,4 +201,34 @@ describe('sampleStore (feature on):', () => {
...
+    .then(() =>
+      redisClient.hgetallAsync('samsto:aspect:___aspect2'))
+    .then((aspect) => {
+      sampleStore.arrayStringsToJson(aspect,
+        sampleStore.constants.fieldsToStringify.aspect);
+      expect(aspect.writers.length).to.equal(1);
+      expect(aspect.writers).to.have
+        .members([user4.name]);
+    })
+    .then(() => samstoinit.init())
```
Comment (annyhe) [inline tests/cache/sampleStore.js]: what is the reason for initializing the sample store at the end of the test?

File: tests/cache/models/aspects/deleteWriters.js [hunk 7]
```diff
@@ -0,0 +1,290 @@
...
+        expect(res.body).to.have.length(2);
+      })
+      .end((_err /* , res */) => {
+        if (_err) {
+          return done(_err);
+        }
+
+        return done();
+      });
+        return null;
```
Comment (annyhe) [inline tests/cache/models/aspects/deleteWriters.js]: is there a reason for returning null? do we need a return here at all?

Comment (shriramshankar) [inline tests/cache/models/aspects/deleteWriters.js]: @annyhe Will correct it, we do not need it.

File: api/v1/controllers/aspects.js (hunk 5 above)
Comment (pallavi2209) [inline api/v1/controllers/aspects.js]: I just thought its better to add comments whenever we are adding a chunk of code, on why/what new code is. But you can ignore this if you feel this is unnecessary.

**Comment by pallavi2209** [comment]
@shriramshankar Just to confirm, my understanding is that I am not required to review this PR.

**Comment by shriramshankar** [comment]
I have resolved the conflicts, added a few more tests and some comments. Can we merge this, please?

**Comment by iamigo** [comment]
ok by me, need's @pallavi2209 to approve the changes since she requested changes

**Comment by shriramshankar** [comment]
@iamigo I added the third reviewer by mistake and there is nothing I can do to remove the third reviewer. If I start working on my bulkUpsert story without this in the master, I will have to resolve the conflict again. Can we please merge this?

**Pull request merged**